### PR TITLE
king dawn: replace web3 usage with hand rolled jsonrpc messages.

### LINKED ISSUE #3370: king: `new` isn't working due to dawn differences.
I tried to create a comet on a droplet, and it failed with "ResponseTimeout".  In `run` I can add `--no-port-forwarding` which should fix this, but I can't seem to add that for `new`.

### PR DESCRIPTION
**Pull request by eglaysher**
This replaces the autogenerated bindings to the Azimuth contracts
which use Network.Web3 with hand rolled json messages. Booting a
ship involved 256 individual galaxy point lookups using web3, while
Vere batched all of that into one JSONRPC message.

With this patch, we also batch everything at each phase into one
JSONRPC batch.

Closes #3370

### CODE REVIEW

**Review by jtobin — CHANGES_REQUESTED** [review]
Batching requests and axing `web3`? #winning 

Most of my commentary here is around general Haskell idioms and style; I'd consider removing the `error` usage in pure code and adding the missing type signatures, in particular.  The Ethereum-related parsing logic looks correct, although I didn't check the gruesome details for every field.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs [hunk 1]
```diff
@@ -74,85 +76,262 @@ ringToPass Ring{..} = Pass{..}
...
+dawnPostRequests :: forall req e resp
+                  . (ToJSON req, RequestMethod req)
+                 => String
+                 -> (req -> Text -> resp)
+                 -> [req]
+                 -> RIO e [resp]
+dawnPostRequests endpoint responseBuilder requests = do
+  -- Encode our input requests
+  let requestPayload =
+        encode $ Array $ fromList $ map toFullRequest $ zip [0..] requests
```
[the remark this answers is not in the document]
Comment (eglaysher) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs:141]: (Changed all `++` to `<>`, and some of the maps to fmaps, depending where in the chain they were.)

File: pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs [hunk 2]
```diff
@@ -6,51 +6,45 @@ module Urbit.Vere.Dawn where
...
+azimuthAddr = "0x223c067f8cf28ae173ee5cafea60ca44c335fecb" :: Text
 
-toBloq :: Quantity -> Bloq
-toBloq = fromIntegral . unQuantity
+-- Conversion Utilities --------------------------------------------------------
 
-passFromEth :: BytesN 32 -> BytesN 32 -> UIntN 32 -> Pass
-passFromEth enc aut sut | sut /= 1 =
+passFromBS :: ByteString -> ByteString -> ByteString -> Pass
+passFromBS enc aut sut | bytesAtom sut /= 1 =
```
[the remark this answers is not in the document]
Comment (eglaysher) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs:39]: This is much better.

File: pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs [hunk 3]
```diff
@@ -74,85 +76,262 @@ ringToPass Ring{..} = Pass{..}
...
+
+  -- Build the final result structure by calling the passed in builder with the
+  -- request (some outputs need data from the request sturcture, eitherwise,
+  -- we'd lean on FromJSON).
+  let results = map (uncurry responseBuilder) (zip requests rawSorted)
+  pure results
+
+ where
+  toFullRequest :: (Int, req) -> Value
+  toFullRequest (id, req) = object [ "jsonrpc" .= ("2.0" :: Text)
```
[the remark this answers is not in the document]
Comment (eglaysher) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs:161]: The parens are actually necessary.

**Review by jtobin — APPROVED** [review]
:+1: :+1:

File: pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs [hunk 4]
```diff
@@ -74,85 +76,262 @@ ringToPass Ring{..} = Pass{..}
...
+
+  ship = Ship $ fromIntegral grqPointId
+  rift = fromIntegral $ bytesAtom rawContinuityNum
+  life = fromIntegral $ bytesAtom rawKeyRevision
+  pass = passFromBS rawEncryptionKey rawAuthenticationKey rawCryptoSuite
+
+removePrefix :: ByteString -> ByteString
+removePrefix withOhEx
+  | prefix == "0x" = suffix
+  | otherwise      = error "not prefixed with 0x"
```
[the remark this answers is not in the document]
Comment (jtobin) [reply pkg/hs/urbit-king/lib/Urbit/Vere/Dawn.hs:289]: :+1: works for me.

**Comment by jtobin** [comment]
Post-review caution: looks like a test broke after limiting the exports.

**Pull request merged**